chore: merge develop into switch PR

Ad96el opened the pull request:
Please start the review at the second commit:

- This commit is the merge of develop into the switch pallet PR: 0897dd24f83bbbfb86fb91f6fcd1f42715207240
- This commit contains all changes required to make the switch pallet compile and updates the XCM version to 4: 073f2d52d2c8fc512017a232a53b1a2dabc0e948
- This commit contains basic renaming of some variables. Since the traits for xcm are using the latest version, I changed the implementation also to the latest: [3fc1277](https://github.com/KILTprotocol/kilt-node/pull/679/commits/3fc1277e7fbdfa26e97c1d20313389c5b16b74cc)

ntn-x2 reviewed (changes requested):
I only focused on the switch pallet and the relevant logic in the runtime.

I tried to point out all places where we use `xcm::latest`, and make sure we use explicit versions, in this case v4. It's a bit like using the `latest` tag in Docker: could go right, but could go wrong if something changes in a new version, and we have no way to know. Hence switching to the new version should be a planned decision.

Other than that, there's few things with the test cases, but big big kudos for having the patience to change them. I'm thinking it might be time to introduce some macros to reduce the boilerplate by a lot.

But great job so far!

Ad96el commented on pallets/pallet-asset-switch/src/xcm/transact/mock.rs:
```diff
@@ -72,19 +75,26 @@ impl frame_system::Config for MockRuntime {
 }
 
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, MaxEncodedLen, Encode, Decode, Debug, TypeInfo, Default)]
+pub struct MockRuntimeHoldReaseon([u8; 4]);
```
I looked in the required trait bounds for RuntimeHoldReaseon. I guess there is more required. In addition, RuntimeHoldReason is requiring VariantCount which is not implemented in an array

Ad96el commented on pallets/pallet-asset-switch/src/xcm/transact/mock.rs:
```diff
@@ -72,19 +75,26 @@ impl frame_system::Config for MockRuntime {
 	type RuntimeEvent = RuntimeEvent;
-	type RuntimeHoldReason = [u8; 4];
+	type RuntimeHoldReason = MockRuntimeHoldReaseon;
+	type RuntimeFreezeReason = ();
```
Ah yes sure. I was thinking we need a proper HoldReason. I removed it in d5968e9438b35d7a51e4dab1f81e286debe8e00b

Ad96el commented on pallets/pallet-asset-switch/src/benchmarking.rs:
```diff
@@ -69,26 +69,29 @@ mod benchmarks {
-	const fn default_info() -> BenchmarkInfo {
-		const DEFAULT_RESERVE_LOCATION: MultiLocation = MultiLocation {
+	fn default_info() -> BenchmarkInfo {
+		let default_reserve_location: Location = Location {
```
Yes. I either need to call an `.into()` or create a new Arc by `Arc::new` which is both not allowed

Ad96el commented on pallets/pallet-asset-switch/src/xcm/trade/xcm_fee_asset/tests/buy_weight.rs:
```diff
@@ -285,8 +372,9 @@ fn successful_on_stored_fungible_xcm_fee_asset_v2_with_input_fungible() {
 		interior: xcm::v2::Junctions::X1(xcm::v2::Junction::Parachain(1_000)),
 	};
 	let new_switch_pair_info = {
+		let location_v3: xcm::v3::MultiLocation = location.try_into().unwrap();
```
Well, the `get_switch_pair_info_for_remote_location` requires a v4 location. Either I have to convert this way or convert it into a `VersionedLocation` and parse that after, or is there a smooth way to convert it directly?

Ad96el commented on pallets/pallet-asset-switch/src/xcm/match/tests.rs:
```diff
@@ -85,15 +85,44 @@ fn successful_with_stored_v3() {
 		parents: 1,
 		interior: xcm::v2::Junctions::X1(xcm::v2::Junction::Parachain(1_000)),
 	};
+	let location_v3: xcm::v3::MultiLocation = location.try_into().unwrap();
```
No, why? I transformed the v2 location to a v3 location. I explained the comment further blow, why I did that.

Ad96el commented:
Yeah, well, it was a quite active design decision to change the specific xcm version to latest for the xcm structs and stay on v4 for the tests. The traits themselves are not versioned and use the latest as well. I thought that for future xcm upgrades, the unit tests and integration tests would reveal any breaking changes. I guess both approaches are fine, so I changed them to version 4. No need for further discussion.

I have addressed all requested changes. It is a bit hard to keep track of the comments, so I did not link the corresponding PR to each comment.

I kindly ask you to review the additional commits made after your last comment.
